src/token/token.service.ts

```diff
@@ -0,0 +1,41 @@
+import { Injectable } from '@nestjs/common';
+import { TokenParamsDto } from './dto/token.params.dto';
+import { ProtocolException } from 'protocol-common/protocol.exception';
+import { ProtocolErrorCode } from 'protocol-common/protocol.errorcode';
+import * as jwt from 'jsonwebtoken';
+import { IJwksService } from '../remote/jwks.service.interface';
+import { JsonWebTokenError } from 'jsonwebtoken';
+
+@Injectable()
+export class TokenService {
+
+    constructor(private readonly jwksService: IJwksService) {}
+
+    /**
+     * Verify the signature on the token and check for an id for the agent within the token.
+     */
+    public async verify(params: TokenParamsDto): Promise<string> {
+        try {
+            // Decode & Verify the token
+            const key = await this.jwksService.getKey(params.token);
+            const pubKey: string = IJwksService.isCertSigningKey(key) ? key.publicKey : key.rsaPublicKey;
+            const token: any = jwt.verify(params.token, new Buffer(pubKey, 'base64'), {
+                algorithms: [process.env.AUTH0_ALGORITHM as any],
+                complete: true
+            });
+
+            // Verify the token's content actually contains an id for the agent
```

Comment:
Style thing: my vote would be for lines 27-32 to be outside the try-catch block so we don't throw and then immediately catch and throw again, but it's not important

Reply:
Let me see if there's a cleaner way to do this... putting it outside is also problematic because then the token could be uninitialized or nullish.